### PR TITLE
Label values with matchers by intersecting postings

### PR DESCRIPTION
**Pull request by colega**
Instead of iterating all matched series to find the values, this checks if each one of the label values is present in the matched series (postings). This is especially faster when matchers match a lot of series but there are actually few values. 

The benchmark results are:

```
name                                                                   old time/op    new time/op      delta
Querier/Head/labelValuesWithMatchers/i_with_n="1"                         157ms ± 0%        48ms ± 0%   ~     (p=1.000 n=1+1)
Querier/Head/labelValuesWithMatchers/i_with_n="^.+$"                      1.80s ± 0%       0.46s ± 0%   ~     (p=1.000 n=1+1)
Querier/Head/labelValuesWithMatchers/i_with_n="1",j!="foo"                144ms ± 0%        57ms ± 0%   ~     (p=1.000 n=1+1)
Querier/Head/labelValuesWithMatchers/i_with_n="1",i=~"^.*$",j!="foo"      304ms ± 0%       111ms ± 0%   ~     (p=1.000 n=1+1)
Querier/Head/labelValuesWithMatchers/n_with_j!="foo"                      761ms ± 0%       164ms ± 0%   ~     (p=1.000 n=1+1)
Querier/Head/labelValuesWithMatchers/n_with_i="1"                        6.11µs ± 0%      6.62µs ± 0%   ~     (p=1.000 n=1+1)
Querier/Block/labelValuesWithMatchers/i_with_n="1"                        117ms ± 0%        62ms ± 0%   ~     (p=1.000 n=1+1)
Querier/Block/labelValuesWithMatchers/i_with_n="^.+$"                     1.44s ± 0%       0.24s ± 0%   ~     (p=1.000 n=1+1)
Querier/Block/labelValuesWithMatchers/i_with_n="1",j!="foo"              92.1ms ± 0%      70.3ms ± 0%   ~     (p=1.000 n=1+1)
Querier/Block/labelValuesWithMatchers/i_with_n="1",i=~"^.*$",j!="foo"     196ms ± 0%       115ms ± 0%   ~     (p=1.000 n=1+1)
Querier/Block/labelValuesWithMatchers/n_with_j!="foo"                     1.23s ± 0%       0.21s ± 0%   ~     (p=1.000 n=1+1)
Querier/Block/labelValuesWithMatchers/n_with_i="1"                       1.06ms ± 0%      0.88ms ± 0%   ~     (p=1.000 n=1+1)

name                                                                   old alloc/op   new alloc/op     delta
Querier/Head/labelValuesWithMatchers/i_with_n="1"                        29.5MB ± 0%      26.9MB ± 0%   ~     (p=1.000 n=1+1)
Querier/Head/labelValuesWithMatchers/i_with_n="^.+$"                     46.8MB ± 0%     251.5MB ± 0%   ~     (p=1.000 n=1+1)
Querier/Head/labelValuesWithMatchers/i_with_n="1",j!="foo"               29.5MB ± 0%      22.3MB ± 0%   ~     (p=1.000 n=1+1)
Querier/Head/labelValuesWithMatchers/i_with_n="1",i=~"^.*$",j!="foo"     46.8MB ± 0%      23.9MB ± 0%   ~     (p=1.000 n=1+1)
Querier/Head/labelValuesWithMatchers/n_with_j!="foo"                     10.3kB ± 0%  138535.2kB ± 0%   ~     (p=1.000 n=1+1)
Querier/Head/labelValuesWithMatchers/n_with_i="1"                        5.54kB ± 0%      7.09kB ± 0%   ~     (p=1.000 n=1+1)
Querier/Block/labelValuesWithMatchers/i_with_n="1"                       39.1MB ± 0%      28.5MB ± 0%   ~     (p=1.000 n=1+1)
Querier/Block/labelValuesWithMatchers/i_with_n="^.+$"                     287MB ± 0%       253MB ± 0%   ~     (p=1.000 n=1+1)
Querier/Block/labelValuesWithMatchers/i_with_n="1",j!="foo"              34.3MB ± 0%      23.9MB ± 0%   ~     (p=1.000 n=1+1)
Querier/Block/labelValuesWithMatchers/i_with_n="1",i=~"^.*$",j!="foo"    51.6MB ± 0%      25.5MB ± 0%   ~     (p=1.000 n=1+1)
Querier/Block/labelValuesWithMatchers/n_with_j!="foo"                     144MB ± 0%       139MB ± 0%   ~     (p=1.000 n=1+1)
Querier/Block/labelValuesWithMatchers/n_with_i="1"                       6.43kB ± 0%      8.66kB ± 0%   ~     (p=1.000 n=1+1)

name                                                                   old allocs/op  new allocs/op    delta
Querier/Head/labelValuesWithMatchers/i_with_n="1"                          104k ± 0%        500k ± 0%   ~     (p=1.000 n=1+1)
Querier/Head/labelValuesWithMatchers/i_with_n="^.+$"                       204k ± 0%        600k ± 0%   ~     (p=1.000 n=1+1)
Querier/Head/labelValuesWithMatchers/i_with_n="1",j!="foo"                 104k ± 0%        500k ± 0%   ~     (p=1.000 n=1+1)
Querier/Head/labelValuesWithMatchers/i_with_n="1",i=~"^.*$",j!="foo"       204k ± 0%        500k ± 0%   ~     (p=1.000 n=1+1)
Querier/Head/labelValuesWithMatchers/n_with_j!="foo"                       66.0 ± 0%       255.0 ± 0%   ~     (p=1.000 n=1+1)
Querier/Head/labelValuesWithMatchers/n_with_i="1"                          61.0 ± 0%       205.0 ± 0%   ~     (p=1.000 n=1+1)
Querier/Block/labelValuesWithMatchers/i_with_n="1"                         304k ± 0%        600k ± 0%   ~     (p=1.000 n=1+1)
Querier/Block/labelValuesWithMatchers/i_with_n="^.+$"                     5.20M ± 0%       0.70M ± 0%   ~     (p=1.000 n=1+1)
Querier/Block/labelValuesWithMatchers/i_with_n="1",j!="foo"                204k ± 0%        600k ± 0%   ~     (p=1.000 n=1+1)
Querier/Block/labelValuesWithMatchers/i_with_n="1",i=~"^.*$",j!="foo"      304k ± 0%        600k ± 0%   ~     (p=1.000 n=1+1)
Querier/Block/labelValuesWithMatchers/n_with_j!="foo"                     3.00M ± 0%       0.00M ± 0%   ~     (p=1.000 n=1+1)
Querier/Block/labelValuesWithMatchers/n_with_i="1"                         61.0 ± 0%       247.0 ± 0%   ~     (p=1.000 n=1+1)
```

The memory usage is higher as we select more series, that is caused by actually expanding the postings, that is a step back, but even in the worst case where we select 1M series:

```
Querier/Head/labelValuesWithMatchers/n_with_j!="foo"                     10.3kB ± 0%  138535.2kB ± 0%
```

There's still a 4x (760ms->160ms) latency improvement:

```
Querier/Head/labelValuesWithMatchers/n_with_j!="foo"                      761ms ± 0%       164ms ± 0%
```

### CODE REVIEW

File: tsdb/head_read.go [hunk 1]
```diff
@@ -103,11 +103,18 @@ func (h *headIndexReader) LabelNames(matchers ...*labels.Matcher) ([]string, err
...
+	case 0:
+		return index.EmptyPostings(), nil
+	case 1:
+		return h.head.postings.Get(name, values[0]), nil
+	default:
+		res := make([]index.Postings, 0, len(values))
+		for _, value := range values {
+			res = append(res, h.head.postings.Get(name, value))
+		}
+		return index.Merge(res...), nil
```
Comment (colega) [inline tsdb/head_read.go:116]: I changed this because `len() == 1` is now (was it before?) a hot path, and I thought it would worth removing the extra unnecessary allocations, does this need a benchmark too?

**Comment by colega** [comment]
I was worried about having to expand the postings and the peak memory usage there, so added a method to `index` that intersects all those postings using a min heap without having to expand them.

This is how the benchmarks looks like now, comparing the old, the intermediate and the latest versions:

```
name \ time/op                                                         old.txt      intersect.txt    intersect_noexpand.txt
Querier/Head/labelValuesWithMatchers/i_with_n="1"                       157ms ± 0%        48ms ± 0%              110ms ± 0%
Querier/Head/labelValuesWithMatchers/i_with_n="^.+$"                    1.80s ± 0%       0.46s ± 0%              0.18s ± 0%
Querier/Head/labelValuesWithMatchers/i_with_n="1",j!="foo"              144ms ± 0%        57ms ± 0%              125ms ± 0%
Querier/Head/labelValuesWithMatchers/i_with_n="1",i=~"^.*$",j!="foo"    304ms ± 0%       111ms ± 0%              177ms ± 0%
Querier/Head/labelValuesWithMatchers/n_with_j!="foo"                    761ms ± 0%       164ms ± 0%              134ms ± 0%
Querier/Head/labelValuesWithMatchers/n_with_i="1"                      6.11µs ± 0%      6.62µs ± 0%             4.29µs ± 0%
Querier/Block/labelValuesWithMatchers/i_with_n="1"                      117ms ± 0%        62ms ± 0%              120ms ± 0%
Querier/Block/labelValuesWithMatchers/i_with_n="^.+$"                   1.44s ± 0%       0.24s ± 0%              0.15s ± 0%
Querier/Block/labelValuesWithMatchers/i_with_n="1",j!="foo"            92.1ms ± 0%      70.3ms ± 0%            125.4ms ± 0%
Querier/Block/labelValuesWithMatchers/i_with_n="1",i=~"^.*$",j!="foo"   196ms ± 0%       115ms ± 0%              170ms ± 0%
Querier/Block/labelValuesWithMatchers/n_with_j!="foo"                   1.23s ± 0%       0.21s ± 0%              0.14s ± 0%
Querier/Block/labelValuesWithMatchers/n_with_i="1"                     1.06ms ± 0%      0.88ms ± 0%             0.92ms ± 0%

name \ alloc/op                                                        old.txt      intersect.txt    intersect_noexpand.txt
Querier/Head/labelValuesWithMatchers/i_with_n="1"                      29.5MB ± 0%      26.9MB ± 0%             19.1MB ± 0%
Querier/Head/labelValuesWithMatchers/i_with_n="^.+$"                   46.8MB ± 0%     251.5MB ± 0%             36.3MB ± 0%
Querier/Head/labelValuesWithMatchers/i_with_n="1",j!="foo"             29.5MB ± 0%      22.3MB ± 0%             19.1MB ± 0%
Querier/Head/labelValuesWithMatchers/i_with_n="1",i=~"^.*$",j!="foo"   46.8MB ± 0%      23.9MB ± 0%             20.7MB ± 0%
Querier/Head/labelValuesWithMatchers/n_with_j!="foo"                   10.3kB ± 0%  138535.2kB ± 0%              6.4kB ± 0%
Querier/Head/labelValuesWithMatchers/n_with_i="1"                      5.54kB ± 0%      7.09kB ± 0%             4.30kB ± 0%
Querier/Block/labelValuesWithMatchers/i_with_n="1"                     39.1MB ± 0%      28.5MB ± 0%             20.7MB ± 0%
Querier/Block/labelValuesWithMatchers/i_with_n="^.+$"                   287MB ± 0%       253MB ± 0%               38MB ± 0%
Querier/Block/labelValuesWithMatchers/i_with_n="1",j!="foo"            34.3MB ± 0%      23.9MB ± 0%             20.7MB ± 0%
Querier/Block/labelValuesWithMatchers/i_with_n="1",i=~"^.*$",j!="foo"  51.6MB ± 0%      25.5MB ± 0%             22.3MB ± 0%
Querier/Block/labelValuesWithMatchers/n_with_j!="foo"                   144MB ± 0%       139MB ± 0%                0MB ± 0%
Querier/Block/labelValuesWithMatchers/n_with_i="1"                     6.43kB ± 0%      8.66kB ± 0%             5.86kB ± 0%

name \ allocs/op                                                       old.txt      intersect.txt    intersect_noexpand.txt
Querier/Head/labelValuesWithMatchers/i_with_n="1"                        104k ± 0%        500k ± 0%               300k ± 0%
Querier/Head/labelValuesWithMatchers/i_with_n="^.+$"                     204k ± 0%        600k ± 0%               400k ± 0%
Querier/Head/labelValuesWithMatchers/i_with_n="1",j!="foo"               104k ± 0%        500k ± 0%               300k ± 0%
Querier/Head/labelValuesWithMatchers/i_with_n="1",i=~"^.*$",j!="foo"     204k ± 0%        500k ± 0%               300k ± 0%
Querier/Head/labelValuesWithMatchers/n_with_j!="foo"                     66.0 ± 0%       255.0 ± 0%              139.0 ± 0%
Querier/Head/labelValuesWithMatchers/n_with_i="1"                        61.0 ± 0%       205.0 ± 0%               87.0 ± 0%
Querier/Block/labelValuesWithMatchers/i_with_n="1"                       304k ± 0%        600k ± 0%               400k ± 0%
Querier/Block/labelValuesWithMatchers/i_with_n="^.+$"                   5.20M ± 0%       0.70M ± 0%              0.50M ± 0%
Querier/Block/labelValuesWithMatchers/i_with_n="1",j!="foo"              204k ± 0%        600k ± 0%               400k ± 0%
Querier/Block/labelValuesWithMatchers/i_with_n="1",i=~"^.*$",j!="foo"    304k ± 0%        600k ± 0%               400k ± 0%
Querier/Block/labelValuesWithMatchers/n_with_j!="foo"                   3.00M ± 0%       0.00M ± 0%              0.00M ± 0%
Querier/Block/labelValuesWithMatchers/n_with_i="1"                       61.0 ± 0%       247.0 ± 0%              129.0 ± 0%
```

TL;DR: results are comparable until 150ms, but when we reach 1s it's a 10x improvement.
Similar improvements for the memory usage.

**Comment by colega** [comment]
@pstibrany suggested me to run benchmarks with count>1, so I did with count=10 and here are the results (I've skipped the results of the intermediate implementation this time):

```
name                                                                   old time/op    new time/op    delta
Querier/Head/labelValuesWithMatchers/i_with_n="1"                         140ms ± 0%     107ms ± 2%   -23.67%  (p=0.000 n=9+9)
Querier/Head/labelValuesWithMatchers/i_with_n="^.+$"                      1.63s ± 6%     0.18s ± 3%   -89.08%  (p=0.000 n=9+9)
Querier/Head/labelValuesWithMatchers/i_with_n="1",j!="foo"                129ms ± 2%     125ms ± 9%    -2.91%  (p=0.035 n=10+10)
Querier/Head/labelValuesWithMatchers/i_with_n="1",i=~"^.*$",j!="foo"      275ms ± 3%     184ms ± 8%   -33.03%  (p=0.000 n=10+10)
Querier/Head/labelValuesWithMatchers/n_with_j!="foo"                      676ms ± 0%     130ms ± 0%   -80.72%  (p=0.000 n=10+9)
Querier/Head/labelValuesWithMatchers/n_with_i="1"                        5.35µs ± 1%    4.43µs ± 5%   -17.28%  (p=0.000 n=9+10)
Querier/Block/labelValuesWithMatchers/i_with_n="1"                        109ms ± 1%     111ms ± 0%    +2.56%  (p=0.000 n=10+10)
Querier/Block/labelValuesWithMatchers/i_with_n="^.+$"                     1.37s ± 1%     0.15s ± 0%   -89.30%  (p=0.000 n=10+10)
Querier/Block/labelValuesWithMatchers/i_with_n="1",j!="foo"              88.7ms ± 1%   117.9ms ± 1%   +32.83%  (p=0.000 n=9+9)
Querier/Block/labelValuesWithMatchers/i_with_n="1",i=~"^.*$",j!="foo"     177ms ± 1%     163ms ± 1%    -7.92%  (p=0.000 n=9+9)
Querier/Block/labelValuesWithMatchers/n_with_j!="foo"                     1.14s ± 1%     0.14s ± 2%   -87.85%  (p=0.000 n=10+10)
Querier/Block/labelValuesWithMatchers/n_with_i="1"                        973µs ± 6%     928µs ± 3%    -4.68%  (p=0.001 n=9+9)

name                                                                   old alloc/op   new alloc/op   delta
Querier/Head/labelValuesWithMatchers/i_with_n="1"                        29.5MB ± 0%    19.1MB ± 0%   -35.36%  (p=0.000 n=9+10)
Querier/Head/labelValuesWithMatchers/i_with_n="^.+$"                     46.8MB ± 0%    36.3MB ± 0%   -22.30%  (p=0.000 n=10+10)
Querier/Head/labelValuesWithMatchers/i_with_n="1",j!="foo"               29.5MB ± 0%    19.1MB ± 0%   -35.37%  (p=0.000 n=9+10)
Querier/Head/labelValuesWithMatchers/i_with_n="1",i=~"^.*$",j!="foo"     46.8MB ± 0%    20.7MB ± 0%   -55.79%  (p=0.000 n=10+10)
Querier/Head/labelValuesWithMatchers/n_with_j!="foo"                     10.3kB ± 2%     6.4kB ± 0%   -37.99%  (p=0.000 n=10+10)
Querier/Head/labelValuesWithMatchers/n_with_i="1"                        5.54kB ± 0%    4.30kB ± 0%   -22.51%  (p=0.000 n=10+10)
Querier/Block/labelValuesWithMatchers/i_with_n="1"                       39.1MB ± 0%    20.7MB ± 0%   -47.14%  (p=0.000 n=10+10)
Querier/Block/labelValuesWithMatchers/i_with_n="^.+$"                     287MB ± 0%      38MB ± 0%   -86.77%  (p=0.000 n=10+10)
Querier/Block/labelValuesWithMatchers/i_with_n="1",j!="foo"              34.3MB ± 0%    20.7MB ± 0%   -39.75%  (p=0.000 n=9+10)
Querier/Block/labelValuesWithMatchers/i_with_n="1",i=~"^.*$",j!="foo"    51.6MB ± 0%    22.3MB ± 0%   -56.80%  (p=0.000 n=10+10)
Querier/Block/labelValuesWithMatchers/n_with_j!="foo"                     144MB ± 0%       0MB ± 0%   -99.99%  (p=0.000 n=7+10)
Querier/Block/labelValuesWithMatchers/n_with_i="1"                       6.43kB ± 0%    5.86kB ± 0%    -8.83%  (p=0.000 n=10+10)

name                                                                   old allocs/op  new allocs/op  delta
Querier/Head/labelValuesWithMatchers/i_with_n="1"                          104k ± 0%      300k ± 0%  +188.57%  (p=0.000 n=9+10)
Querier/Head/labelValuesWithMatchers/i_with_n="^.+$"                       204k ± 0%      400k ± 0%   +96.11%  (p=0.000 n=10+10)
Querier/Head/labelValuesWithMatchers/i_with_n="1",j!="foo"                 104k ± 0%      300k ± 0%  +188.51%  (p=0.000 n=9+10)
Querier/Head/labelValuesWithMatchers/i_with_n="1",i=~"^.*$",j!="foo"       204k ± 0%      300k ± 0%   +47.06%  (p=0.000 n=10+10)
Querier/Head/labelValuesWithMatchers/n_with_j!="foo"                       66.5 ± 1%     139.0 ± 0%  +109.02%  (p=0.000 n=10+10)
Querier/Head/labelValuesWithMatchers/n_with_i="1"                          61.0 ± 0%      87.0 ± 0%   +42.62%  (p=0.000 n=10+10)
Querier/Block/labelValuesWithMatchers/i_with_n="1"                         304k ± 0%      400k ± 0%   +31.60%  (p=0.000 n=10+10)
Querier/Block/labelValuesWithMatchers/i_with_n="^.+$"                     5.20M ± 0%     0.50M ± 0%   -90.39%  (p=0.000 n=10+10)
Querier/Block/labelValuesWithMatchers/i_with_n="1",j!="foo"                204k ± 0%      400k ± 0%   +96.11%  (p=0.000 n=9+10)
Querier/Block/labelValuesWithMatchers/i_with_n="1",i=~"^.*$",j!="foo"      304k ± 0%      400k ± 0%   +31.58%  (p=0.000 n=10+10)
Querier/Block/labelValuesWithMatchers/n_with_j!="foo"                     3.00M ± 0%     0.00M ± 0%   -99.99%  (p=0.000 n=7+10)
Querier/Block/labelValuesWithMatchers/n_with_i="1"                         61.0 ± 0%     129.0 ± 0%  +111.48%  (p=0.000 n=10+10)
```

File: tsdb/head_read.go (hunk 1 above)
Comment (codesome) [reply tsdb/head_read.go:116]: No need of benchmark for this. It is only removing stuff for this case and not changing anything.

File: tsdb/index/postings.go [hunk 2]
```diff
@@ -831,3 +833,80 @@ type seriesRefSlice []storage.SeriesRef
...
+// FindIntersectingPostings checks the intersection of p and candidates[i] for each i in candidates,
+// if intersection is non empty, then i is added to the indexes returned.
+// Returned indexes are not sorted.
+func FindIntersectingPostings(p Postings, candidates []Postings) (indexes []int, err error) {
+	h := make(postingsWithIndexHeap, 0, len(candidates))
+	for idx, it := range candidates {
+		if it.Next() {
+			h = append(h, postingsWithIndex{index: idx, p: it})
+		} else {
+			if it.Err() != nil {
```
Comment (codesome) [inline tsdb/index/postings.go:846]:
```suggestion
		} else if it.Err() != nil {
```

File: tsdb/index/postings_test.go [hunk 3]
```diff
@@ -929,3 +932,142 @@ func TestMemPostings_Delete(t *testing.T) {
...
+	t.Run("multiple intersections", func(t *testing.T) {
+		p := NewListPostings([]storage.SeriesRef{10, 15, 20, 25, 30, 35, 40, 45, 50})
+		candidates := []Postings{
+			0: NewListPostings([]storage.SeriesRef{7, 13, 14, 27}), // does not intersect
+			1: NewListPostings([]storage.SeriesRef{10, 20}),        // does intersect
+			2: NewListPostings([]storage.SeriesRef{29, 30, 31}),    // does intersect
+			3: NewListPostings([]storage.SeriesRef{29, 30, 31}),    // does intersect (same again)
+			4: NewListPostings([]storage.SeriesRef{60}),            // does not intersect
+			5: NewListPostings([]storage.SeriesRef{45}),            // does intersect
+			6: EmptyPostings(),                                     // does not intersect
```
Comment (codesome) [inline tsdb/index/postings_test.go:946]:
```suggestion
			0: NewListPostings([]storage.SeriesRef{7, 13, 14, 27}), // Does not intersect.
			1: NewListPostings([]storage.SeriesRef{10, 20}),        // Does intersect.
			2: NewListPostings([]storage.SeriesRef{29, 30, 31}),    // Does intersect.
			3: NewListPostings([]storage.SeriesRef{29, 30, 31}),    // Does intersect (same again).
			4: NewListPostings([]storage.SeriesRef{60}),            // Does not intersect.
			5: NewListPostings([]storage.SeriesRef{45}),            // Does intersect.
			6: EmptyPostings(),                                     // Does not intersect.
```

File: tsdb/index/postings_test.go [hunk 4]
```diff
@@ -929,3 +932,142 @@ func TestMemPostings_Delete(t *testing.T) {
...
+		sort.Ints(indexes)
+		require.Equal(t, []int{1, 2, 3, 5}, indexes)
+	})
+
+	t.Run("no intersections", func(t *testing.T) {
+		p := NewListPostings([]storage.SeriesRef{10, 15, 20, 25, 30, 35, 40, 45, 50})
+		candidates := []Postings{
+			0: NewListPostings([]storage.SeriesRef{7, 13, 14, 27}), // does not intersect
+			1: NewListPostings([]storage.SeriesRef{60}),            // does not intersect
+			2: EmptyPostings(),                                     // does not intersect
```
Comment (codesome) [inline tsdb/index/postings_test.go:960]:
```suggestion
			0: NewListPostings([]storage.SeriesRef{7, 13, 14, 27}), // Does not intersect.
			1: NewListPostings([]storage.SeriesRef{60}),            // Does not intersect.
			2: EmptyPostings(),                                     // Does not intersect.
```

File: tsdb/querier.go [hunk 5]
```diff
@@ -375,38 +375,31 @@ func inversePostingsForMatcher(ix IndexReader, m *labels.Matcher) (index.Posting
...
-		dedupe[v] = nil
 	}
-
-	if err = p.Err(); err != nil {
-		return nil, err
+	indexes, err := index.FindIntersectingPostings(p, valuesPostings)
+	if err != nil {
+		return nil, errors.Wrap(err, "intersecting postings")
 	}
+	sort.Ints(indexes)
```
Comment (codesome) [inline tsdb/querier.go:398]: We don't need it sorted (we were not sorting `values` before)

File: tsdb/querier_bench_test.go [hunk 6]
```diff
@@ -143,6 +153,38 @@ func benchmarkPostingsForMatchers(b *testing.B, ir IndexReader) {
...
+	jNotFoo := labels.MustNewMatcher(labels.MatchNotEqual, "j", "foo")
+	n1 := labels.MustNewMatcher(labels.MatchEqual, "n", "1"+postingsBenchSuffix)
+	nPlus := labels.MustNewMatcher(labels.MatchRegexp, "i", "^.+$")
+
+	cases := []struct {
+		name      string
+		labelName string
+		matchers  []*labels.Matcher
+	}{
+		// i has 100k values
```
Comment (codesome) [inline tsdb/querier_bench_test.go:168]:
```suggestion
		// i has 100k values.
```

File: tsdb/querier_bench_test.go [hunk 7]
```diff
@@ -143,6 +153,38 @@ func benchmarkPostingsForMatchers(b *testing.B, ir IndexReader) {
...
+		{`i with n="1"`, "i", []*labels.Matcher{n1}},
+		{`i with n="^.+$"`, "i", []*labels.Matcher{nPlus}},
+		{`i with n="1",j!="foo"`, "i", []*labels.Matcher{n1, jNotFoo}},
+		{`i with n="1",i=~"^.*$",j!="foo"`, "i", []*labels.Matcher{n1, iStar, jNotFoo}},
+		// n has 10 values
```
Comment (codesome) [inline tsdb/querier_bench_test.go:173]:
```suggestion
		// n has 10 values.
```

**Review by codesome — COMMENTED** [review]
Almost LGTM, some nits, thanks! I talked with @colega offline and the increase in allocs seems to be side effect of `NewListPostings` calls through the benchmarks and not really a concern (we can always revert this back if it gives problems later)

File: tsdb/index/postings.go (hunk 2 above)
Comment (colega) [reply tsdb/index/postings.go]: Done in https://github.com/prometheus/prometheus/pull/9907/commits/26023f79bf8b36bf530eff2b0019c99908d1c429

File: tsdb/querier.go (hunk 5 above)
Comment (colega) [reply tsdb/querier.go]: Oh right, removed in https://github.com/prometheus/prometheus/pull/9907/commits/b0143cb602b315dafbfa464199fc007971727a00

**Review by codesome — APPROVED** [review]

**Comment by colega** [comment]
@codesome yes, most of the allocations I've seen were coming from the NewListPostings called for every postings call for each label value, so with 100k label cardinality that causes 100k allocations. 

The results are mainly same magnitude as before, and I think that the 2-3x allocations increase is paid off by the memory and cpu improvements.

And yes, we can always revert this, or maybe just add a `sync.Pool` for `index.ListPostings` instance and dispose them after using.

**Comment by colega** [comment]
I digged a little bit more into profiles for both [head](https://share.polarsignals.com/e0625df/) and [block](https://share.polarsignals.com/3f4c6ef/) cases, and focusing on the `labelValuesWithMatchers` they look like this:

### Head 

image

### Block

image

And I noticed that actually big part of allocations happens in the `Pop` method of the heap implementation:

https://github.com/prometheus/prometheus/blob/3947238ce0e3ff7e7c6e9502f73275cdbc233054/tsdb/index/postings.go#L904-L910

Then I realized that this allocation is just because the return type is an `interface{}`. Just for fun, I decided to compare how it would look with generics, and [wrote an experimental commit](https://github.com/colega/prometheus/commit/c05b3c3ca74fb84dadae93b1896114f5e8bb2015)

And yes, using generics there reduces the allocations by 25% (for the test case `BenchmarkQuerier/Head/labelValuesWithMatchers/i_with_n="1"$` that goes from 400k to 300k, because since there are 100k label values, we allocate 100k times less when calling Pop())

So, I decided to see if there's any way to avoid that allocation (by avoiding the Pop call) and it seems that there's one! PR pending.
